In [1]:
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [2]:
id = pd.read_csv('insurance_data.csv')      # Reading insurance data file
ed = pd.read_csv('employee_data.csv')       # Reading employee data file
vd = pd.read_csv('vendor_data.csv')         # Reading vendor data file

In [3]:
id.head()   # Insurance data

TXN_DATE_TIME TRANSACTION_ID CUSTOMER_ID POLICY_NUMBER POLICY_EFF_DT  \
0  2020-06-01 00:00:00    TXN00000001   A00003822   PLC00008468    2015-06-23   
1  2020-06-01 00:00:00    TXN00000002   A00008149   PLC00009594    2018-04-21   
2  2020-06-01 00:00:00    TXN00000003   A00003172   PLC00007969    2019-10-03   
3  2020-06-01 00:00:00    TXN00000004   A00007572   PLC00009292    2016-11-29   
4  2020-06-01 00:00:00    TXN00000005   A00008173   PLC00000204    2011-12-26   

      LOSS_DT   REPORT_DT INSURANCE_TYPE  PREMIUM_AMOUNT  CLAIM_AMOUNT  \
0  2020-05-16  2020-05-21         Health          157.13          9000   
1  2020-05-13  2020-05-18       Property          141.71         26000   
2  2020-05-21  2020-05-26       Property          157.24         13000   
3  2020-05-14  2020-05-19         Health          172.87         16000   
4  2020-05-17  2020-05-22         Travel           88.53          3000   

          CUSTOMER_NAME              ADDRESS_LINE1 ADDRESS_LINE2         CITY  \
0  Christopher Demarest  7701 West Saint John Road         #2010     Glendale   
1        Ricardo Gatlin       8595 West 81st Drive           NaN       Arvada   
2        Lashawn Engles        637 Britannia Drive           NaN      Vallejo   
3        Steven Bassett           2803 River Drive           NaN  Thunderbolt   
4       Jason Rodriguez        7573 National Drive           NaN    Livermore   

  STATE  POSTAL_CODE          SSN MARITAL_STATUS  AGE  TENURE  \
0    AZ        85308  087-11-1946              Y   54      89   
1    CO        80005  685-33-3536              N   61      80   
2    CA        94591  378-36-0672              N   47      68   
3    GA        31404  669-92-1861              Y   36      16   
4    CA        94550  703-40-1033              Y   51      16   

  EMPLOYMENT_STATUS  NO_OF_FAMILY_MEMBERS RISK_SEGMENTATION HOUSE_TYPE  \
0                 Y                     3                 L        Own   
1                 Y                     4                 L       Rent   
2                 Y                     6                 L       Rent   
3                 Y                     7                 L   Mortgage   
4                 Y                     2                 M       Rent   

  SOCIAL_CLASS  ROUTING_NUMBER         ACCT_NUMBER CUSTOMER_EDUCATION_LEVEL  \
0           LI       109134974  HXJP58258181908465                 Bachelor   
1           MI        40125819  JUND46859540983731                 Bachelor   
2           MI        99513168  WGZZ90128415227650                      PhD   
3           MI        18429110  WIKE91555436351397                  Masters   
4           HI        70752391  VYJW71311537294027                  Masters   

  CLAIM_STATUS INCIDENT_SEVERITY AUTHORITY_CONTACTED  ANY_INJURY  \
0            A        Major Loss              Police           0   
1            A        Total Loss           Ambulance           1   
2            A        Total Loss              Police           0   
3            A        Minor Loss           Ambulance           0   
4            A        Major Loss              Police           0   

   POLICE_REPORT_AVAILABLE INCIDENT_STATE   INCIDENT_CITY  \
0                        1             GA        Savannah   
1                        0             AL      Montgomery   
2                        1             CO  Grand Junction   
3                        0             GA        Savannah   
4                        1             TN       Nashville   

   INCIDENT_HOUR_OF_THE_DAY    AGENT_ID  VENDOR_ID  
0                         4  AGENT00413  VNDR00556  
1                         0  AGENT00769  VNDR00592  
2                        19  AGENT00883  VNDR00031  
3                        12  AGENT00278  VNDR00075  
4                        18  AGENT00636  VNDR00472

In [4]:
id.shape       

(10000, 38)

In [5]:
ed.head()      # Employee data

AGENT_ID      AGENT_NAME DATE_OF_JOINING          ADDRESS_LINE1  \
0  AGENT00001       Ray Johns      1993-06-05       1402 Maggies Way   
1  AGENT00002   Angelo Borjon      2005-12-27         414 Tanya Pass   
2  AGENT00003  Candy Spellman      2003-09-02    606 National Street   
3  AGENT00004      Mary Smith      2004-09-23  235 Hugh Thomas Drive   
4  AGENT00005    Mildred Diaz      2011-06-21  3426 Broadview Street   

  ADDRESS_LINE2              CITY STATE  POSTAL_CODE  EMP_ROUTING_NUMBER  \
0           NaN  Waterbury Center    VT         5677            34584958   
1           NaN       Panama City    FL        32404           107363763   
2          #306      Fayetteville    AR        72701            81744097   
3           NaN       Panama City    FL        32404            67563771   
4           NaN        Montgomery    AL        36110           114951317   

      EMP_ACCT_NUMBER  
0  HKUN51252328472585  
1  OPIS19290040088204  
2  YSCJ67489688482590  
3  ZANG21285355574581  
4  DZFS82244494451134

In [6]:
ed.shape

(1200, 10)

In [7]:
vd.head()  # Vendor data

VENDOR_ID              VENDOR_NAME                   ADDRESS_LINE1  \
0  VNDR00001  King, Proctor and Jones        2027 North Shannon Drive   
1  VNDR00002               Garcia Ltd          5701 East Shirley Lane   
2  VNDR00003               Cherry LLC            1217 Cottondale Road   
3  VNDR00004              Mays-Benson  227 West Montgomery Cross Road   
4  VNDR00005               Wilson PLC            23 North Hill Street   

  ADDRESS_LINE2          CITY STATE  POSTAL_CODE  
0            #5  Fayetteville    AR        72703  
1           NaN    Montgomery    AL        36117  
2           NaN    Montgomery    AL        36109  
3          #736      Savannah    GA        31406  
4           NaN     Nashville    TN        37210

In [8]:
vd.shape

(600, 7)

## Task 1 : Merge the 3 dataset and create 1 view of data. You can merge insurance_data.csv and employee_data.csv on AGENT_ID You can merge insurance_data.csv and vendor_data.csv on VENDOR_ID Note: Use left Outer join as not all claims require Vendor

In [9]:
## Merging dataframes with appropriate suffixes for future reference

df1 = pd.merge(id,ed,on='AGENT_ID',how='left',suffixes=('_CUSTOMER','_AGENT'))
view = pd.merge(df1,vd,on='VENDOR_ID',how='left')

In [10]:
view.head()

TXN_DATE_TIME TRANSACTION_ID CUSTOMER_ID POLICY_NUMBER POLICY_EFF_DT  \
0  2020-06-01 00:00:00    TXN00000001   A00003822   PLC00008468    2015-06-23   
1  2020-06-01 00:00:00    TXN00000002   A00008149   PLC00009594    2018-04-21   
2  2020-06-01 00:00:00    TXN00000003   A00003172   PLC00007969    2019-10-03   
3  2020-06-01 00:00:00    TXN00000004   A00007572   PLC00009292    2016-11-29   
4  2020-06-01 00:00:00    TXN00000005   A00008173   PLC00000204    2011-12-26   

      LOSS_DT   REPORT_DT INSURANCE_TYPE  PREMIUM_AMOUNT  CLAIM_AMOUNT  \
0  2020-05-16  2020-05-21         Health          157.13          9000   
1  2020-05-13  2020-05-18       Property          141.71         26000   
2  2020-05-21  2020-05-26       Property          157.24         13000   
3  2020-05-14  2020-05-19         Health          172.87         16000   
4  2020-05-17  2020-05-22         Travel           88.53          3000   

          CUSTOMER_NAME     ADDRESS_LINE1_CUSTOMER ADDRESS_LINE2_CUSTOMER  \
0  Christopher Demarest  7701 West Saint John Road                  #2010   
1        Ricardo Gatlin       8595 West 81st Drive                    NaN   
2        Lashawn Engles        637 Britannia Drive                    NaN   
3        Steven Bassett           2803 River Drive                    NaN   
4       Jason Rodriguez        7573 National Drive                    NaN   

  CITY_CUSTOMER STATE_CUSTOMER  POSTAL_CODE_CUSTOMER          SSN  \
0      Glendale             AZ                 85308  087-11-1946   
1        Arvada             CO                 80005  685-33-3536   
2       Vallejo             CA                 94591  378-36-0672   
3   Thunderbolt             GA                 31404  669-92-1861   
4     Livermore             CA                 94550  703-40-1033   

  MARITAL_STATUS  AGE  TENURE EMPLOYMENT_STATUS  NO_OF_FAMILY_MEMBERS  \
0              Y   54      89                 Y                     3   
1              N   61      80                 Y                     4   
2              N   47      68                 Y                     6   
3              Y   36      16                 Y                     7   
4              Y   51      16                 Y                     2   

  RISK_SEGMENTATION HOUSE_TYPE SOCIAL_CLASS  ROUTING_NUMBER  \
0                 L        Own           LI       109134974   
1                 L       Rent           MI        40125819   
2                 L       Rent           MI        99513168   
3                 L   Mortgage           MI        18429110   
4                 M       Rent           HI        70752391   

          ACCT_NUMBER CUSTOMER_EDUCATION_LEVEL CLAIM_STATUS INCIDENT_SEVERITY  \
0  HXJP58258181908465                 Bachelor            A        Major Loss   
1  JUND46859540983731                 Bachelor            A        Total Loss   
2  WGZZ90128415227650                      PhD            A        Total Loss   
3  WIKE91555436351397                  Masters            A        Minor Loss   
4  VYJW71311537294027                  Masters            A        Major Loss   

  AUTHORITY_CONTACTED  ANY_INJURY  POLICE_REPORT_AVAILABLE INCIDENT_STATE  \
0              Police           0                        1             GA   
1           Ambulance           1                        0             AL   
2              Police           0                        1             CO   
3           Ambulance           0                        0             GA   
4              Police           0                        1             TN   

    INCIDENT_CITY  INCIDENT_HOUR_OF_THE_DAY    AGENT_ID  VENDOR_ID  \
0        Savannah                         4  AGENT00413  VNDR00556   
1      Montgomery                         0  AGENT00769  VNDR00592   
2  Grand Junction                        19  AGENT00883  VNDR00031   
3        Savannah                        12  AGENT00278  VNDR00075   
4       Nashville                        18  AGENT00636  VNDR00472   

        AG

In [11]:
view.shape

(10000, 53)

## Task 2 : Business Leader wants to find Top 3 Insurance Type where we are getting most insurance claims?

In [12]:
id.INSURANCE_TYPE.value_counts()[:3]

Property    1692
Mobile      1692
Health      1690
Name: INSURANCE_TYPE, dtype: int64

## Task 3 : Business Leader wants to find Top 5 States where we are getting most insurance claims for customer belonging to HIGH(H) risk segment?

In [13]:
id.columns

Index(['TXN_DATE_TIME', 'TRANSACTION_ID', 'CUSTOMER_ID', 'POLICY_NUMBER',
       'POLICY_EFF_DT', 'LOSS_DT', 'REPORT_DT', 'INSURANCE_TYPE',
       'PREMIUM_AMOUNT', 'CLAIM_AMOUNT', 'CUSTOMER_NAME', 'ADDRESS_LINE1',
       'ADDRESS_LINE2', 'CITY', 'STATE', 'POSTAL_CODE', 'SSN',
       'MARITAL_STATUS', 'AGE', 'TENURE', 'EMPLOYMENT_STATUS',
       'NO_OF_FAMILY_MEMBERS', 'RISK_SEGMENTATION', 'HOUSE_TYPE',
       'SOCIAL_CLASS', 'ROUTING_NUMBER', 'ACCT_NUMBER',
       'CUSTOMER_EDUCATION_LEVEL', 'CLAIM_STATUS', 'INCIDENT_SEVERITY',
       'AUTHORITY_CONTACTED', 'ANY_INJURY', 'POLICE_REPORT_AVAILABLE',
       'INCIDENT_STATE', 'INCIDENT_CITY', 'INCIDENT_HOUR_OF_THE_DAY',
       'AGENT_ID', 'VENDOR_ID'],
      dtype='object')

In [14]:
"""Top 5 States where we are getting most insurance claims for customers belonging to HIGH(H) risk segment are:"""

id[id['RISK_SEGMENTATION']=='H'].STATE.value_counts()[:5]

CA    148
AZ    105
FL    104
TN    100
AR     97
Name: STATE, dtype: int64

## Task 4 : Business wants to create a new variable “COLOCATION” which will have following values IF Customer State == Incident State == Agent Address State THEN 1 ELSE 0 Find the mean of this new column

In [15]:
"""The columns above in view with no suffixes represent column of vendor data"""

view['COLOCATION'] = (view['STATE_CUSTOMER']==view['INCIDENT_STATE'])&(view['INCIDENT_STATE']==view['STATE_AGENT'])

In [16]:
""" Changing column values from True to 1 and False to 0 """

view['COLOCATION']=view.COLOCATION.astype('int')

In [17]:
view['COLOCATION'].mean()

0.0044

##### The mean of COLOCATION column is 0.0044

## Task 5 : Data entry error was detected in the data and you are required to correct it. If for any claim transaction “AUTHORITY_CONTACTED” is NOT “Police” and POLICE_AVAILABLE == 1 Then Update “AUTHORITY_CONTACTED” to Police.

In [18]:
non_police = view[view['AUTHORITY_CONTACTED']!='Police']

In [19]:
update_rows = non_police[non_police['POLICE_REPORT_AVAILABLE']==1].index

In [20]:
## Updating the Authority column
view.loc[update_rows,'AUTHORITY_CONTACTED'] = 'Police'

In [21]:
# Checking
view[(view['POLICE_REPORT_AVAILABLE']==1)&(view['AUTHORITY_CONTACTED']!='Police')]

Empty DataFrame
Columns: [TXN_DATE_TIME, TRANSACTION_ID, CUSTOMER_ID, POLICY_NUMBER, POLICY_EFF_DT, LOSS_DT, REPORT_DT, INSURANCE_TYPE, PREMIUM_AMOUNT, CLAIM_AMOUNT, CUSTOMER_NAME, ADDRESS_LINE1_CUSTOMER, ADDRESS_LINE2_CUSTOMER, CITY_CUSTOMER, STATE_CUSTOMER, POSTAL_CODE_CUSTOMER, SSN, MARITAL_STATUS, AGE, TENURE, EMPLOYMENT_STATUS, NO_OF_FAMILY_MEMBERS, RISK_SEGMENTATION, HOUSE_TYPE, SOCIAL_CLASS, ROUTING_NUMBER, ACCT_NUMBER, CUSTOMER_EDUCATION_LEVEL, CLAIM_STATUS, INCIDENT_SEVERITY, AUTHORITY_CONTACTED, ANY_INJURY, POLICE_REPORT_AVAILABLE, INCIDENT_STATE, INCIDENT_CITY, INCIDENT_HOUR_OF_THE_DAY, AGENT_ID, VENDOR_ID, AGENT_NAME, DATE_OF_JOINING, ADDRESS_LINE1_AGENT, ADDRESS_LINE2_AGENT, CITY_AGENT, STATE_AGENT, POSTAL_CODE_AGENT, EMP_ROUTING_NUMBER, EMP_ACCT_NUMBER, VENDOR_NAME, ADDRESS_LINE1, ADDRESS_LINE2, CITY, STATE, POSTAL_CODE, COLOCATION]
Index: []

## Task 6 : Business wants to check the Claim Amount for deviation for each transaction, they would like you to calculate as follow CLAIM_DEVIATION = AVG_CLAIM_AMOUNT_FOR_LAST_30DAYS (same insurance type) / CURRENT_CLAIM_AMOUNT. If the value < 0.5 THEN CLAIM_DEVIATION = 1 ELSE 0. If there is less than 30 days of transaction history THEN -1 Note: LAST_30DAYS does not include current day

In [22]:
view.TXN_DATE_TIME.max()

'2021-06-30 00:00:00'

In [23]:
# Assuming current day to be 2021-06-30
curr_date = pd.to_datetime('2021-06-30')

In [24]:
view['TXN_DATE_TIME'] = pd.to_datetime(view['TXN_DATE_TIME'])

In [25]:
## df with last 30 days data
last_30 = view.loc[(view['TXN_DATE_TIME'] > '2021-05-30') & (view['TXN_DATE_TIME'] < '2021-06-30')]  

In [26]:
# Getting average claim of last 30 days insurance type wise and storing it in dictionary
average_claim = last_30.groupby('INSURANCE_TYPE').mean()['CLAIM_AMOUNT'].to_dict()

In [27]:
# defining a function for calculating claim deviation
def Claim_dev(row):
    if (curr_date - row['TXN_DATE_TIME'])<pd.Timedelta(days=30):
        return -1
    claim_deviation = average_claim[row['INSURANCE_TYPE']] / row['CLAIM_AMOUNT']
    if claim_deviation < 0.5:
        return 1
    else:
        return 0

In [28]:
# Applying the function to the dataframe
view['CLAIM_DEVIATION'] = view.apply(Claim_dev,axis=1)

In [29]:
view.CLAIM_DEVIATION.value_counts()

 0    9216
-1     784
Name: CLAIM_DEVIATION, dtype: int64

There are 9216 claims with deviation less than 0.5

784 claims with transaction history of less than 30 days

## Task 7 : Find All Agents who have worked on more than 2 types of Insurance Claims. Sort them by Total Claim Amount Approved under them in descending order

In [30]:
"""Calculating insurance types worked on of each agent"""

type_count = view.groupby('AGENT_ID')['INSURANCE_TYPE'].nunique()

In [31]:
"""Calculating whether agent has more than 2 insurance types"""

agent_insurance_types = type_count>2

In [32]:
"""Getting list of agents who have worked on more than 2 types of insurance types"""

multi_type_agents = agent_insurance_types[agent_insurance_types].index

In [33]:
Top_agents = view.groupby(['AGENT_ID','AGENT_NAME']).sum().loc[multi_type_agents][['CLAIM_AMOUNT']].sort_values('CLAIM_AMOUNT',ascending=False)

Top_agents

CLAIM_AMOUNT
AGENT_ID   AGENT_NAME                        
AGENT00807 Don Filkins                 528800
AGENT00679 Clara Barnett               489000
AGENT00771 Roger Burns                 422100
AGENT00125 Salvador Soriano            400400
AGENT00789 Alison Baron                392900
AGENT00525 Don Ritchie                 385900
AGENT00319 Frances Kimble              375600
AGENT00388 Larry Scott                 373800
AGENT00439 Nancy Hannah                370000
AGENT00482 Jose Dame                   368000
AGENT00881 Maria Middleton             366400
AGENT00763 Paul Banerjee               361000
AGENT00914 John Wilson                 360000
AGENT00219 Gary Grigsby                359500
AGENT00729 Samuel Tolman               358100
AGENT01094 Robert Venable              351000
AGENT01142 Frank Kuss                  350000
AGENT00743 Roberto Wetzel              339200
AGENT00920 Yolanda Samples             335700
AGENT00875 Phillip Moore               332000
AGENT00313 Jean Cedeno                 331000
AGENT01174 Trina Hill                  327100
AGENT00182 Arlene Lopez                324500
AGENT00633 Clarence Mcclendon          322000
AGENT00986 Christopher Lazor           318100
AGENT00943 Jessica Perkins             314300
AGENT00123 Tanya Fontenot              313500
AGENT00968 Kathryn Lehman              313400
AGENT00312 Danielle Caddell            312900
AGENT00158 Jim Rippee                  311000
AGENT00290 Marjorie Roberts            310800
AGENT00806 Martin Eaton                309000
AGENT00814 Kelli Miles                 306000
AGENT00307 John Roberts                304300
AGENT00302 Brenda Trent                299200
AGENT00534 Richard Sheehan             298500
AGENT00004 Mary Smith                  297600
AGENT01101 Andre Gault                 297200
AGENT00509 Ryan Otoole                 297200
AGENT00038 Harold Ward                 296000
AGENT00254 Freda Dudley                295600
AGENT00083 Christopher Fontenot        295400
AGENT00837 John Colson                 295300
AGENT00576 Bertha Reynolds             294500
AGENT00664 Howard Planas               294300
AGENT00849 Debbra Collins              293800
AGENT01175 Nidia Snider                289000
AGENT00838 Elbert Duncan               288700
AGENT01162 James Martinez              288500
AGENT01158 Paul Hanson                 288000
AGENT00200 Kayla Place                 286200
AGENT00992 James Velasco               285600
AGENT00589 Jeffrey Jackson             282400
AGENT00521 Sarah Gabriel               282000
AGENT00446 Christina Bock              282000
AGENT01127 Robert Ward                 280200
AGENT00260 Cleveland Feyh              279000
AGENT00250 Adrian Calderon             278600
AGENT00460 Jamie Williams              278400
AGENT00111 Philip Mabrey               278000
AGENT00646 Emma Tubbs                  276000
AGENT00713 Ashley Jackson              276000
AGENT00612 Susan Mincey                276000
AGENT00607 Joyce Bumgardner            275600
AGENT00598 Alicia Workman              275200
AGENT01079 Willie Issa                 275200
AGENT01196 Myrtle Mccatty              274000
AGENT00308 Ernesto Brown               272500
AGENT01135 Shannon Kee                 272000
AGENT00486 Ray Higdon                  272000
AGENT00492 James Wood                  271500
AGENT00944 Melvin Stacey               270100
AGENT01117 Karen Rawling               270000
AGENT00406 John Hill                   269900
AGENT00611 Michael Olson               269400
AGENT00177 Fran Maney                  269000
AGENT00069 Christian Nusbaum           268900
AGENT00737 Joan Schuldt                268100
AGENT00391 Bernice Mcneeley            267900
AGENT01027 Peter Bates                 266700
AGENT00309 Lee Nieland                 266300
AGENT00241 Kevin Howard                266000
AGENT00367 Nancy Marbley               264200
AGENT00491 Joel Nix                    263400
AGENT00397 Humberto Tanguay            263200
AGENT00688 Glenda Briney       

In [34]:
''' Agents who have worked on more than 2 types of insurance claims, sorted by total claims approved under them in 
    descending order along with types of insurance they worked in '''

Top_agents.join(type_count)

CLAIM_AMOUNT  INSURANCE_TYPE
AGENT_ID   AGENT_NAME                                        
AGENT00807 Don Filkins                 528800               6
AGENT00679 Clara Barnett               489000               4
AGENT00771 Roger Burns                 422100               5
AGENT00125 Salvador Soriano            400400               4
AGENT00789 Alison Baron                392900               4
AGENT00525 Don Ritchie                 385900               6
AGENT00319 Frances Kimble              375600               6
AGENT00388 Larry Scott                 373800               6
AGENT00439 Nancy Hannah                370000               5
AGENT00482 Jose Dame                   368000               4
AGENT00881 Maria Middleton             366400               6
AGENT00763 Paul Banerjee               361000               5
AGENT00914 John Wilson                 360000               3
AGENT00219 Gary Grigsby                359500               5
AGENT00729 Samuel Tolman               358100               5
AGENT01094 Robert Venable              351000               3
AGENT01142 Frank Kuss                  350000               5
AGENT00743 Roberto Wetzel              339200               5
AGENT00920 Yolanda Samples             335700               6
AGENT00875 Phillip Moore               332000               4
AGENT00313 Jean Cedeno                 331000               4
AGENT01174 Trina Hill                  327100               5
AGENT00182 Arlene Lopez                324500               5
AGENT00633 Clarence Mcclendon          322000               5
AGENT00986 Christopher Lazor           318100               4
AGENT00943 Jessica Perkins             314300               5
AGENT00123 Tanya Fontenot              313500               5
AGENT00968 Kathryn Lehman              313400               6
AGENT00312 Danielle Caddell            312900               5
AGENT00158 Jim Rippee                  311000               6
AGENT00290 Marjorie Roberts            310800               5
AGENT00806 Martin Eaton                309000               3
AGENT00814 Kelli Miles                 306000               4
AGENT00307 John Roberts                304300               5
AGENT00302 Brenda Trent                299200               6
AGENT00534 Richard Sheehan             298500               6
AGENT00004 Mary Smith                  297600               6
AGENT01101 Andre Gault                 297200               3
AGENT00509 Ryan Otoole                 297200               4
AGENT00038 Harold Ward                 296000               4
AGENT00254 Freda Dudley                295600               5
AGENT00083 Christopher Fontenot        295400               6
AGENT00837 John Colson                 295300               5
AGENT00576 Bertha Reynolds             294500               5
AGENT00664 Howard Planas               294300               6
AGENT00849 Debbra Collins              293800               6
AGENT01175 Nidia Snider                289000               4
AGENT00838 Elbert Duncan               288700               4
AGENT01162 James Martinez              288500               5
AGENT01158 Paul Hanson                 288000               5
AGENT00200 Kayla Place                 286200               5
AGENT00992 James Velasco               285600               6
AGENT00589 Jeffrey Jackson             282400               6
AGENT00521 Sarah Gabriel               282000               5
AGENT00446 Christina Bock              282000               4
AGENT01127 Robert Ward                 280200               5
AGENT00260 Cleveland Feyh              279000               4
AGENT00250 Adrian Calderon             278600               5
AGENT00460 Jamie Williams              278400               6
AGENT00111 Philip Mabrey               278000               6
AGENT00646 Emma Tubbs                  276000               3
AGENT00713 Ashley Jackson              276000               4
AGENT00612 Susan Mincey                276000               4
AGE

## Task 8 : Mobile & Travel Insurance premium are discounted by 10%. Health and Property Insurance premium are increased by 7%. Life and Motor Insurance premium are marginally increased by 2%. What will be overall change in % of the Premium Amount Collected for all these Customer?


In [35]:
premium_collected_earlier = view['PREMIUM_AMOUNT'].sum()

In [36]:
"""Total premium collected by each insurance type"""

prem_by_types = view.groupby('INSURANCE_TYPE').sum()['PREMIUM_AMOUNT'].to_dict()

prem_by_types

{'Health': 252455.73,
 'Life': 125621.22,
 'Mobile': 15191.42,
 'Motor': 165391.98,
 'Property': 202285.83,
 'Travel': 124139.77}

In [37]:
""" Mobile & Travel premium discounted by 10% """
prem_by_types['Mobile']*=0.9
prem_by_types['Travel']*=0.9

""" Health and Property premium increases by 7%  """
prem_by_types['Health']*=1.07
prem_by_types['Property']*=1.07

""" Life and Motor premium increased by 2% """
prem_by_types['Life']*=1.02
prem_by_types['Motor']*=1.02

In [38]:
""" Premium after discount and increase """
revised_premium = sum(prem_by_types.values())

In [39]:
print("premium_collected_earlier = ", premium_collected_earlier,'\n',"revised_premium = ",revised_premium)

premium_collected_earlier =  885085.95 
 revised_premium =  908805.0042000001


In [40]:
(revised_premium-premium_collected_earlier)/premium_collected_earlier *100

2.6798588543858517

#### Overall change in % of the Premium Amount is 2.68 % 

## Task 9 : Business wants to give discount to customer who are loyal and under stress due to Covid 19. They have laid down an eligibility Criteria as follow IF CUSTOMER_TENURE > 60 AND EMPLOYMENT_STATUS = “N” AND NO_OF_FAMILY_MEMBERS >=4 THEN 1 ELSE 0 Create a new column “ELIGIBLE_FOR_DISCOUNT” and find it mean.

In [41]:
""" Calculating status for discount with given conditions """

disc_status=(view['TENURE']>60)&(view['EMPLOYMENT_STATUS']=='N')&(view['NO_OF_FAMILY_MEMBERS']>=4)

In [42]:
""" Creating an column ELIGIBLE_FOR_DISCOUNT with above values """

view['ELIGIBLE_FOR_DISCOUNT'] = disc_status.astype('int')

In [43]:
view['ELIGIBLE_FOR_DISCOUNT'].value_counts()

0    9701
1     299
Name: ELIGIBLE_FOR_DISCOUNT, dtype: int64

In [44]:
""" Mean of this column """

view.ELIGIBLE_FOR_DISCOUNT.mean()

0.0299

## Task 10 : Business wants to check Claim Velocity which is defined as follow CLAIM_VELOCITY = NO_OF_CLAIMS_IN_LAST30DAYS (for the current insurance type)/ NO_OF_CLAIMS_IN_LAST3DAYS (for the current insurance type) 
### Note: LAST30DAYS & LAST3DAYS does not include current day

In [45]:
curr_date

Timestamp('2021-06-30 00:00:00')

In [46]:
view['CLAIM_STATUS'].value_counts()

A    9497
D     503
Name: CLAIM_STATUS, dtype: int64

In [47]:
## We had defined last 30 days data in task 6 earlier, similarly defining for last 3 days

In [48]:
last_3_days = view.loc[(view['TXN_DATE_TIME'] > curr_date-pd.Timedelta(days=4)) & (view['TXN_DATE_TIME'] < curr_date)]

##---> pd.Timedelta is set to 4 days because it also counts current date

In [49]:
## Calculating total approved claims insurance type wise in last 30 days

claims_in_last_30_days = pd.DataFrame.from_dict(last_30[last_30['CLAIM_STATUS']=='A'].groupby('INSURANCE_TYPE').size().to_dict(),orient='index',columns=['NO_OF_CLAIMS_IN_LAST30DAYS'])

In [50]:
## Calculating total approved claims insurance type wise in last 3 days

claims_in_last_3_days = pd.DataFrame.from_dict(last_3_days[last_3_days['CLAIM_STATUS']=='A'].groupby('INSURANCE_TYPE').size().to_dict(),orient='index',columns=['NO_OF_CLAIMS_IN_LAST3DAYS'])

In [51]:
approved_claims = claims_in_last_30_days.join(claims_in_last_3_days)

In [52]:
approved_claims['CLAIM_VELOCITY'] = approved_claims['NO_OF_CLAIMS_IN_LAST30DAYS']/approved_claims['NO_OF_CLAIMS_IN_LAST3DAYS']

In [53]:
approved_claims

NO_OF_CLAIMS_IN_LAST30DAYS  NO_OF_CLAIMS_IN_LAST3DAYS  \
Health                           128                         14   
Life                             128                         12   
Mobile                           127                          9   
Motor                            117                         12   
Property                         125                         12   
Travel                           123                         14   

          CLAIM_VELOCITY  
Health          9.142857  
Life           10.666667  
Mobile         14.111111  
Motor           9.750000  
Property       10.416667  
Travel          8.785714

##### Claim velocity is calculated for only claims approved, similarly ot is calculated for all claims below.

In [54]:
All_claims30D = pd.DataFrame.from_dict(last_30.groupby('INSURANCE_TYPE').size().to_dict(),orient='index',columns=['NO_OF_CLAIMS_IN_LAST30DAYS'])

In [55]:
All_claims3D = pd.DataFrame.from_dict(last_3_days.groupby('INSURANCE_TYPE').size().to_dict(),orient='index',columns=['NO_OF_CLAIMS_IN_LAST3DAYS'])

In [56]:
total_claims=All_claims30D.join(All_claims3D)

In [57]:
total_claims['CLAIM_VELOCITY'] = total_claims['NO_OF_CLAIMS_IN_LAST30DAYS']/total_claims['NO_OF_CLAIMS_IN_LAST3DAYS']

In [58]:
total_claims

NO_OF_CLAIMS_IN_LAST30DAYS  NO_OF_CLAIMS_IN_LAST3DAYS  \
Health                           136                         14   
Life                             140                         14   
Mobile                           134                         10   
Motor                            122                         13   
Property                         129                         13   
Travel                           125                         14   

          CLAIM_VELOCITY  
Health          9.714286  
Life           10.000000  
Mobile         13.400000  
Motor           9.384615  
Property        9.923077  
Travel          8.928571

## Task 11 : Find all low performing agents i.e. employees who are in the bottom 5 percentile based on Claims worked by them.


In [59]:
bottom_5percentile_limit = view.groupby('AGENT_ID').sum().quantile(0.05)['CLAIM_AMOUNT']

In [60]:
Agent_grouped_data = view.groupby('AGENT_ID').sum()

In [61]:
## Low performing agent data
low_performing_agents = Agent_grouped_data[Agent_grouped_data['CLAIM_AMOUNT']<=bottom_5percentile_limit][['CLAIM_AMOUNT']]

In [62]:
low_performing_agents

CLAIM_AMOUNT
AGENT_ID                
AGENT00006         22400
AGENT00014          7400
AGENT00016         27000
AGENT00085         25700
AGENT00088         27700
AGENT00091         30000
AGENT00098         28200
AGENT00134         22000
AGENT00149         28900
AGENT00156         13200
AGENT00172         22200
AGENT00185         16100
AGENT00186         27700
AGENT00188         28400
AGENT00197         18500
AGENT00252         20000
AGENT00263         11500
AGENT00275         16200
AGENT00279         28000
AGENT00300         12600
AGENT00310         28800
AGENT00316         22200
AGENT00351         21700
AGENT00425         17800
AGENT00442         30300
AGENT00536         20000
AGENT00562         21100
AGENT00571         24200
AGENT00585         16600
AGENT00590         22200
AGENT00604          5500
AGENT00621          5300
AGENT00628          6100
AGENT00672         21300
AGENT00689         13700
AGENT00706         10800
AGENT00721          9000
AGENT00732         11300
AGENT00773         15000
AGENT00780         11300
AGENT00790          6000
AGENT00804         20000
AGENT00805         22600
AGENT00830         23500
AGENT00841         17800
AGENT00856         24200
AGENT00894         27600
AGENT00910         15500
AGENT00969         28300
AGENT00985         30000
AGENT01021         24300
AGENT01037          5600
AGENT01052         28000
AGENT01077          4200
AGENT01107         24700
AGENT01128         28000
AGENT01131         27500
AGENT01151         28000
AGENT01154           800
AGENT01179         27300

## Task 12 :  Business wants to find all Suspicious Employees (Agents). IF TOTAL CLAIM AMOUNT which meet below criteria is >= 15000 THEN AGENT is classified as Suspicious ELSE Not
### CLAIM_STATUS = Approved AND CUSTOMER_RISK_SEGMENTATION = High AND INCIDENT_SEVERITY = “Major Loss”
#### If Suspicious, then 1 ELSE 0. Find mean of this column.

In [63]:
criteria = (view['CLAIM_STATUS']=='A')&(view['RISK_SEGMENTATION']=='H')&(view['INCIDENT_SEVERITY']=='Major Loss')

In [64]:
### Transactions that follow criteria

criteria_met = view[criteria]

In [65]:
grouped_suspicion_data = criteria_met.groupby('AGENT_ID').sum()

In [66]:
### Storing agents_id that are suspicious(met all the criteria)

suspicious_agents = list(grouped_suspicion_data[grouped_suspicion_data['CLAIM_AMOUNT']>=15000].index)

In [67]:
ed.head()

AGENT_ID      AGENT_NAME DATE_OF_JOINING          ADDRESS_LINE1  \
0  AGENT00001       Ray Johns      1993-06-05       1402 Maggies Way   
1  AGENT00002   Angelo Borjon      2005-12-27         414 Tanya Pass   
2  AGENT00003  Candy Spellman      2003-09-02    606 National Street   
3  AGENT00004      Mary Smith      2004-09-23  235 Hugh Thomas Drive   
4  AGENT00005    Mildred Diaz      2011-06-21  3426 Broadview Street   

  ADDRESS_LINE2              CITY STATE  POSTAL_CODE  EMP_ROUTING_NUMBER  \
0           NaN  Waterbury Center    VT         5677            34584958   
1           NaN       Panama City    FL        32404           107363763   
2          #306      Fayetteville    AR        72701            81744097   
3           NaN       Panama City    FL        32404            67563771   
4           NaN        Montgomery    AL        36110           114951317   

      EMP_ACCT_NUMBER  
0  HKUN51252328472585  
1  OPIS19290040088204  
2  YSCJ67489688482590  
3  ZANG21285355574581  
4  DZFS82244494451134

In [68]:
ed['SUSPICIOUS'] = 0

In [69]:
sus_emp_slice = ed['AGENT_ID'].isin(suspicious_agents)

In [70]:
#### Changing suspicious column value to 1 of suspicious employees

ed.loc[sus_emp_slice,'SUSPICIOUS']=1

In [71]:
ed['SUSPICIOUS'].mean()

0.12416666666666666

##### Ans - The mean of suspicious column is 0.124166 that means from all employees 12.42% are suspicious